### PR TITLE
[Feature] Add PoseWarper dataset

### PR DESCRIPTION
**Pull request by liqikai9**
## Motivation

Add [PoseWarper](https://arxiv.org/abs/1906.04016) dataset part. See also #910.

## Modification

1. Add a base class in `mmpose/datasets/datasets/base/kpt_2d_sview_rgb_img_top_down_dataset.py`.

Add PoseWarper dataset in `/mmpose/datasets/datasets/top_down/topdown_posetrack18video_dataset.py`

2. Modify 
`mmpose/datasets/__init__.py` ,
`mmpose/datasets/datasets/__init__.py` ,
 `mmpose/datasets/datasets/top_down/__init__.py` ,
to add the new class.

3. Modify related pipelines, including

- `mmpose/datasets/pipelines/loading.py` to support loading multiple images for an instance, modify the `LoadImageFromFile` class.
-
- `mmpose/datasets/pipelines/top_down_transform.py` to support transforming multiple images,  `TopDownRandomFlip`, `TopDownAffine`,

- `mmpose/datasets/pipelines/shared_transform.py` to support loading a list of images.

4. Add unittests for the above modifications, including

- `tests/test_backward_compatibility/test_dataset_info_compatibility/test_top_down_dataset_compatibility.py`,

- `tests/test_datasets/test_top_down_dataset.py`

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/datasets/datasets/__init__.py [hunk 1]
```diff
@@ -8,36 +8,35 @@
 from .bottom_up import (BottomUpAicDataset, BottomUpCocoDataset,
                         BottomUpCocoWholeBodyDataset, BottomUpCrowdPoseDataset,
                         BottomUpMhpDataset)
-from .face import (Face300WDataset, FaceAFLWDataset, FaceCocoWholeBodyDataset,
```
Comment (jin-s13) [inline mmpose/datasets/datasets/__init__.py:11]: `FaceCocoWholeBodyDataset` is mistakenly deleted.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+                'nframes': nframes,
+                'frame_id': frame_id,
+                'frame_weight': self.frame_weight
+            })
+            bbox_id = bbox_id + 1
+        print(f'=> Total boxes after filter '
+              f'low score@{self.det_bbox_thr}: {bbox_id}')
+        return kpt_db
+
+    def evaluate(self, outputs, res_folder, metric='mAP', **kwargs):
```
Comment (jin-s13) [inline mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:334]: This part is not covered by unit-test

File: mmpose/datasets/datasets/__init__.py (hunk 1 above)
Comment (liqikai9) [reply mmpose/datasets/datasets/__init__.py:11]: OK, I will add it.

File: mmpose/datasets/datasets/__init__.py [hunk 3]
```diff
@@ -8,36 +8,35 @@
...
-                   FaceCOFWDataset, FaceWFLWDataset)
+from .face import (Face300WDataset, FaceAFLWDataset, FaceCOFWDataset,
+                   FaceWFLWDataset)
 from .fashion import DeepFashionDataset
-from .hand import (FreiHandDataset, HandCocoWholeBodyDataset,
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply mmpose/datasets/datasets/__init__.py:14]: OK, I will add it.

File: mmpose/datasets/datasets/__init__.py [hunk 4]
```diff
@@ -8,36 +8,35 @@
...
-                   InterHand2DDataset, InterHand3DDataset, OneHand10KDataset,
-                   PanopticDataset)
+from .hand import (FreiHandDataset, InterHand2DDataset, InterHand3DDataset,
+                   OneHand10KDataset, PanopticDataset)
 from .mesh import (MeshAdversarialDataset, MeshH36MDataset, MeshMixDataset,
                    MoshDataset)
 from .top_down import (TopDownAicDataset, TopDownCocoDataset,
                        TopDownCocoWholeBodyDataset, TopDownCrowdPoseDataset,
-                       TopDownH36MDataset, TopDownHalpeDataset,
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply mmpose/datasets/datasets/__init__.py:21]: OK, I will add it.

File: mmpose/datasets/datasets/top_down/__init__.py [hunk 5]
```diff
@@ -4,18 +4,18 @@
...
 from .topdown_ochuman_dataset import TopDownOCHumanDataset
 from .topdown_posetrack18_dataset import TopDownPoseTrack18Dataset
+from .topdown_posetrack18_video_dataset import TopDownPoseTrack18VideoDataset
 
 __all__ = [
     'TopDownAicDataset', 'TopDownCocoDataset', 'TopDownCocoWholeBodyDataset',
     'TopDownCrowdPoseDataset', 'TopDownMpiiDataset', 'TopDownMpiiTrbDataset',
     'TopDownOCHumanDataset', 'TopDownPoseTrack18Dataset',
-    'TopDownJhmdbDataset', 'TopDownMhpDataset', 'TopDownH36MDataset',
-    'TopDownHalpeDataset'
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply mmpose/datasets/datasets/top_down/__init__.py:20]: OK, I will add it.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+            cur_image_file = os.path.join(self.img_prefix,
+                                          self.id2name[img_id])
+            image_files.append(cur_image_file)
+
+            prev_nm = file_name.split('/')[-1]
+            ref_idx = int(prev_nm.replace('.jpg', ''))
+
+            # select the frame indices
+            if not self.test_mode and self.frame_indexes_train is not None:
+                indexes = self.frame_indexes_train
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:199]: OK, I will change it.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+            joints_3d_visible[:, :2] = np.minimum(1, keypoints[:, 2:3])
+
+            center, scale = self._xywh2cs(*obj['clean_bbox'][:4])
+
+            image_files = []
+            cur_image_file = os.path.join(self.img_prefix,
+                                          self.id2name[img_id])
+            image_files.append(cur_image_file)
+
+            prev_nm = file_name.split('/')[-1]
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:194]: I will use `prev_name` instead.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+            elif not self.test_mode and self.frame_index_rand:
+                low, high = self.frame_index_range
+                indexes = np.random.randint(low, high + 1, self.num_adj_frames)
+            else:
+                indexes = self.frame_indexes_test
+
+            for index in indexes:
+                if self.test_mode and index == 0:
+                    continue
+                sup_idx = ref_idx + index
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:209]: It means the **supporting** frame here.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+                sup_idx = np.clip(sup_idx, 0, nframes - 1)
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:210]: The clip operation is kept as the source code did. I think it was intended for the case when the required frame didn't exist.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py (hunk 2 above)
Comment (liqikai9) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:334]: I will add the corresponding unittest.

File: setup.cfg [hunk 10]
```diff
@@ -19,6 +19,6 @@ line_length = 79
 multi_line_output = 0
 known_standard_library = pkg_resources,setuptools
 known_first_party = mmpose
-known_third_party = PIL,cv2,h5py,json_tricks,matplotlib,mmcv,munkres,numpy,pytest,pytorch_sphinx_theme,requests,scipy,seaborn,spacepy,titlecase,torch,torchvision,xmltodict,xtcocotools
+known_third_party = PIL,cv2,h5py,json_tricks,matplotlib,mmcv,munkres,numpy,poseval,pytest,pytorch_sphinx_theme,requests,scipy,seaborn,spacepy,titlecase,torch,torchvision,xmltodict,xtcocotools
```
[the remark this answers is not in the document]
Comment (liqikai9) [reply setup.cfg:22]: OK, I will change it.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jin-s13) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:209]: I will recommend using `support_idx` instead.

[the remark this answers is not in the document]
Comment (liqikai9) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:209]: OK, I will change it.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py (hunk 2 above)
Comment (liqikai9) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:334]: It seems that when testing the `evaluate` function in both `TopDownPoseTrack18Dataset` and `TopDownPoseTrack18VideoDataset`, we need the corresponding ground truth annotations in folder `tests/data/posetrack/annotations`, which is currently missing.

File: mmpose/datasets/pipelines/loading.py [hunk 11]
```diff
@@ -6,13 +6,6 @@
 
 @PIPELINES.register_module()
 class LoadImageFromFile:
-    """Loading image from file.
```
Comment (luminxu) [inline mmpose/datasets/pipelines/loading.py:9]: why these notes are deleted?

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+            joints_3d_visible[:, :2] = np.minimum(1, keypoints[:, 2:3])
+
+            center, scale = self._xywh2cs(*obj['clean_bbox'][:4])
+
+            image_files = []
+            cur_image_file = os.path.join(self.img_prefix,
+                                          self.id2name[img_id])
+            image_files.append(cur_image_file)
+
+            prev_name = file_name.split('/')[-1]
```
Comment (luminxu) [inline mmpose/datasets/datasets/top_down/topdown_posetrack18_video_dataset.py:199]: rename ``prev_name`` as ``cur_image_name``

**Review by luminxu — APPROVED** [review]

File: mmpose/datasets/pipelines/loading.py (hunk 11 above)
Comment (liqikai9) [reply mmpose/datasets/pipelines/loading.py:9]: These are mistakenly deleted. I will recover them.

**Review by jin-s13 — APPROVED** [review]

**Review by ly015 — APPROVED** [review]

**Pull request merged**